app/src/main/java/com/guzelgimadieva/tasky/core/network/NetworkResponseState.kt

```diff
@@ -0,0 +1,20 @@
+package com.guzelgimadieva.tasky.core.network
+
+sealed interface NetworkResponseState<T> {
+
+    data class Success<T>(val data: T) : NetworkResponseState<T>
+    data class Error<T>(val exception: Exception) : NetworkResponseState<T>
+    fun onSuccess(block: (T) -> Unit): NetworkResponseState<T> {
+        if (this is Success) {
+            block(data)
+        }
+        return this
+    }
+
+    fun onError(block: (Throwable) -> Unit): NetworkResponseState<T> {
+        if (this is Error) {
+            block(exception)
+        }
+        return this
+    }
+}
```

Comment:
For a generic error handling wrapper, I love this version:
```
sealed interface Result<out D, out E: Error> {
    data class Success<out D>(val data: D): Result<D, Nothing>
    data class Error<out E: com.plcoding.tasky.core.domain.util.Error>(val error: E): Result<Nothing, E>
}

inline fun <T, E: Error, R> Result<T, E>.map(map: (T) -> R): Result<R, E> {
    return when(this) {
        is Result.Error -> Result.Error(error)
        is Result.Success -> Result.Success(map(data))
    }
}

fun <T, E: Error> Result<T, E>.asEmptyDataResult(): EmptyResult<E> {
    return map {  }
}

inline fun <T, E: Error> Result<T, E>.onSuccess(action: (T) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> this
        is Result.Success -> {
            action(data)
            this
        }
    }
}
inline fun <T, E: Error> Result<T, E>.onError(action: (E) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> {
            action(error)
            this
        }
        is Result.Success -> this
    }
}

typealias EmptyResult<E> = Result<Unit, E>
```
It handles data and errors in a type-safe way. You'll need a marker interface to mark error enums:
```
interface Error
```
And then you can extend this across the codebase with different error types, e.g.:
```
sealed interface DataError: Error {
    enum class Remote: DataError {
        NOT_FOUND,
        NO_INTERNET,
        UNAUTHORIZED,
        ...
    }
    
    enum class Local: DataError {
        DISK_FULL,
        ...
    }
}
```
Works great:
```
repository.signUp(payload).onSuccess { responseData ->

}.onError { error -> 
    // toUiText() can be an extension of DataError which maps an error to the corresponding message
    _state.update { it.copy(error = error.toUiText()) }
}
```